### PR TITLE
Adding OpenSSL Engine Support

### PR DESCRIPTION
**Pull request by CIPop**

### CODE REVIEW

File: adapters/x509_openssl.c [hunk 1]
```diff
@@ -162,79 +164,158 @@ static int load_key_RSA(SSL_CTX* ssl_ctx, EVP_PKEY* evp_key)
...
+                {
+                    result = 0;
                 }
+
                 EVP_PKEY_free(evp_key);
             }
-            BIO_free(bio_key);
+        }
+
+        if (!ENGINE_finish(engine))  // Releaase functional reference.
```
Comment (danewalton) [inline adapters/x509_openssl.c:264]:
```suggestion
        if (!ENGINE_finish(engine))  // Release functional reference.
```

File: adapters/x509_openssl.c [hunk 2]
```diff
@@ -162,79 +164,158 @@ static int load_key_RSA(SSL_CTX* ssl_ctx, EVP_PKEY* evp_key)
...
+            result = MU_FAILURE;
+        }
+        else
+        {
+            EVP_PKEY* evp_key = ENGINE_load_private_key(engine, x509privatekey_id, NULL, NULL);
 
-                if (result == 0)
+            if (evp_key == NULL)
+            {
+                log_ERR_get_error("failure loading private key from ENGINE");
```
Comment (danewalton) [inline adapters/x509_openssl.c:245]: nit: Maybe try and unify the log messages here? Something like
`Failure loading private key from ENGINE`
`Unable to configure ENGINE`
`Unable to initialize ENGINE`
`Failure loading private key cert from ENGINE`
etc

**Review by danewalton — APPROVED** [review]

File: adapters/tlsio_openssl.c [hunk 3]
```diff
@@ -225,6 +229,34 @@ static void* tlsio_openssl_CloneOption(const char* name, const void* value)
...
+            }
+        }
+        else if (strcmp(name, OPTION_OPENSSL_PRIVATE_KEY_TYPE) == 0)
+        {
+            OPTION_OPENSSL_KEY_TYPE key_type_value = *((OPTION_OPENSSL_KEY_TYPE*)value);
+            OPTION_OPENSSL_KEY_TYPE* value_clone;
+
+            if ((value_clone = (OPTION_OPENSSL_KEY_TYPE*)malloc(sizeof(OPTION_OPENSSL_KEY_TYPE))) == NULL)
+            {
+                LogError("Failed clonning cert or private key option");
```
Comment (ericwol-msft) [inline adapters/tlsio_openssl.c:251]:
```suggestion
                LogError("Failed cloning cert or private key option");
```

**Review by ericwol-msft — APPROVED** [review]

File: adapters/tlsio_openssl.c (hunk 3 above)
Comment (CIPop) [reply adapters/tlsio_openssl.c:251]: Fixed all "clonning" to "cloning".

**Pull request merged**